### ISSUE TITLE
[BUG] vbar_stack with DataFrame as source ignores entire row if first data column is NaN

### ISSUE THREAD

**Issue by pohlt** [bokeh/bokeh#9568]
# Software versions
OS: Linux 5.4.6
Browser: Chrome 78.0.3904.108 (Official Build) (64-bit)
Python: 3.8.1
JupyterLab: 1.2.4
Pandas: 0.25.3
Bokeh: 1.4.0

# Issue
When plotting a `vbar_stack`, the entire row in the data source is ignored if the first input column contains a NaN in a `pandas.DataFrame`.

If a standard Python `dict` is used as a data source, the output is plotted as expected.

```
from bokeh.plotting import figure, show
from bokeh.models import ColumnDataSource

data = dict(index=[1, 2, 3, 4],
            a=[4, 5, None, 7],
            b=[9, None, 7, 6])

source = ColumnDataSource(data)

f = figure()
f.vbar_stack(x="index", stackers=["a", "b"], source=source,
             width=0.5, color=["red", "blue"])
show(f)

source = ColumnDataSource(pd.DataFrame(index=[1, 2, 3, 4],
                                       data=[dict(a=4, b=9),
                                             dict(a=5, b=None),
                                             dict(a=None, b=7),
                                             dict(a=7, b=6),
                                            ]))

f = figure()
f.vbar_stack(x="index", stackers=["a", "b"], source=source,
             width=0.5, color=["red", "blue"])
show(f)
```

## Expected behavior with dict as data source

![Expected: dict as data source](https://user-images.githubusercontent.com/5302291/71999553-eb4a4e80-3241-11ea-90b0-629ac1c0f3bb.png)

## Unexpected behavior with DataFrame as data source

![Unexpected: DataFrame as data source](https://user-images.githubusercontent.com/5302291/71999567-f2715c80-3241-11ea-8f85-cd42e623ecdc.png)

# Mitigation

I currently let Pandas fill the `NaN`s with zeros.

Thanks for Bokeh, guys!

**Comment by bryevdv** [comment]
@pohlt Thanks for the detailed issue and the kind words. Do you have happen to have any comparisons to how other tools behave in this same scenario? I don't think I'd need much convincing that handing nan's explicitly and intentionally in this way is a reasonable thing to do but a few comparisons would be nice to have first.

**Comment by pohlt** [comment]
Good idea. I will check matplotlib and Vega tomorrow for their handling of NaNs, but that's just the consistency with other tools.
More important IMHO is the internal consistency, i.e., that two data sources with equivalent data give the same results and that the position of the NaN (first or later column) doesn't matter.

________________________________

**Comment by bryevdv** [comment]
> More important IMHO is the internal consistency, i.e., that two data sources with equivalent data give the same results and that the position of the NaN (first or later column) doesn't matter.

I agree in principle but there are also some things that are unfortunately outside out control. The difference between the "list" and "DataFrame" versions almost 100% certainly boils down ultimately to the fact that the "list" version is serialized as plain JSON (and the JSON spec inexcusably omits serializing nans or infs 😡), while the DataFrame version gets serialized in a way that can faithfully preserves NaN values all the way in to JS Typed Arrays. We'll certainly do what we an to smooth any inconsistencies, or better document them if that is all that can be done.

**Comment by pohlt** [comment]
## Matplotlib
Does not even try to be clever and just complains that it does not know how to add an `int` and a `None`. Not a bad choice. At least, this makes it obvious that the NaNs are the root problem.

## Vega Lite
Gives the expected result, just as Bokeh does without DataFrames as source:
```
{
  "$schema": "https://vega.github.io/schema/vega-lite/v4.json",
  "data": {"values": [
    {"x": 1, "t": "a", "value":    4}, {"x": 1, "t": "b", "value":    9},
    {"x": 2, "t": "a", "value":    5}, {"x": 2, "t": "b", "value": null},
    {"x": 3, "t": "a", "value": null}, {"x": 3, "t": "b", "value":    7},
    {"x": 4, "t": "a", "value":    7}, {"x": 4, "t": "b", "value":    6}
    ]},
  "mark": "bar",
  "encoding": {
    "x": {
      "field": "x",
      "type": "ordinal"
    },
    "y": {
      "field": "value",
      "type": "quantitative"
    },
    "color": {
      "field": "t",
      "type": "nominal"
    }
  }
}
```
![vega](https://user-images.githubusercontent.com/5302291/72049801-fa6fe180-32bf-11ea-962e-c1838cd35313.png)

Let me point out once more: What bugs me most is the inconsistency how Index 2 and 3 are handled by Bokeh with a DataFrame:
* Index 2 (a=5, b=None): A bar is drawn for a, but none for b. Excellent!
* Index 3 (a=None, b=7): No bar is drawn, although b is 7. Hm, why?
Shouldn't any JSON issues influence both indices in a similar way?

**Comment by bryevdv** [comment]
@pohlt that case ("always in a data frame") I think we can probably make consistent. Earlier I was really referring to the list vs data frame differences.

**Comment by Ak-Bhatia** [comment]
When None is passed inside the Pandas DataFrame, pandas implicitly converts it into NaN, whereas when None is passed into a dictionary, it remains NoneType only. Also, Vbar_stack uses Vbar sequentially on various stackers(a,b,c) and values of those stackers(a,b,c) are sequentially passed onto stack (in case of Vbar_stack, it uses double_stack i.e one for top value and another for bottom value). Top of those stacks is the sum of the elements present in that stack, so when it encounter NaN, it stops further addition and thus no value after that is plotted(can be seen in case of 3 columns). But in case of None value, top of stack keeps on adding values of next columns(because None acts as just a missing value).

**### The one using DataFrame:**
![df](https://user-images.githubusercontent.com/33599912/74103453-1641fe00-4b72-11ea-8561-fc7029986615.png)

**### The one using Dictionary:**
![dict](https://user-images.githubusercontent.com/33599912/74103455-180bc180-4b72-11ea-81d9-a68caf1c5bda.png)

**Comment by Ak-Bhatia** [comment]
When None is passed inside the Pandas DataFrame, pandas implicitly converts it into NaN, whereas when None is passed into a dictionary, it remains NoneType only. Also, Vbar_stack uses Vbar sequentially on various stackers(a,b,c) and values of those stackers(a,b,c) are sequentially passed onto stack (in case of Vbar_stack, it uses double_stack i.e one for top value and another for bottom value). Top of those stacks is the sum of the elements present in that stack, so when it encounter NaN, it stops further addition and thus no value after that is plotted(can be seen in case of 3 columns). But in case of None value, top of stack keeps on adding values of next columns(because None acts as just a missing value).

### The one using DataFrame
![df2](https://user-images.githubusercontent.com/33599912/74103552-f959fa80-4b72-11ea-8aea-11d2d37a8fa5.png)
![df](https://user-images.githubusercontent.com/33599912/74103551-f6f7a080-4b72-11ea-86ad-968f41228566.png)


### The one using Dictionary
![dict2](https://user-images.githubusercontent.com/33599912/74103564-0545bc80-4b73-11ea-9d93-630e6ebd050f.png)
![dict](https://user-images.githubusercontent.com/33599912/74103561-037bf900-4b73-11ea-8055-36f4ad77cec7.png)

**So, the solution for that may be, when the stack encounters any NaN, it should be converted to 0, but with the original dataset to remain same.**

**Comment by bryevdv** [comment]
Changing user data (i.e. converting nan's to zero) might appropriate for a higher level library built on Bokeh, but I do not support that action for Bokeh itself. As it is currently CDS hols a reference to the pandas series, so doing this either means:

* Changing actual user data silently out from under then
* Adding new logging to copy ever pandas series or numpy array instead of just keeping a reference. 

The first option is definitely not good to do as changing data out from under users is generally not a good thing to do. The second options means making every single plotting option substantially more expensive due to data copying. Either way changing data like this may subvert user expectations, or potentially silently hide problems. 

What I would (possibly) support is a change on the JS side to detect nulls in arrays and treat them like NaNs instead. Then consistency would be achieved by having the dict case be the same as the pandas case (no bar at all drawn). Then if users want something different than that it is up to them to explicitly convert Nones or NaNs the way they intend.

**Comment by Hemansh31** [comment]
@bryevdv can I work on this issue. Also, this is my first time ever contributing to an open source code it would be real helpful if you could guide me a little bit :-)

**Comment by bryevdv** [comment]
@Hemansh31 That'd be great, the first step is to be able to build BokehJS locally. Id suggest going through our Developer's Guide "Getting Set Up" chapter:

https://docs.bokeh.org/en/1.4.0/docs/dev_guide/setup.html#devguide-setup

Let us know if you encounter any difficulties along the way.

**Comment by Ak-Bhatia** [comment]
Oh..I wanted to work on this issue..

**Comment by Hemansh31** [comment]
@bryevdv I have gone through this and have already setup everything

**Comment by bryevdv** [comment]
@Ak-Bhatia @Hemansh31 I don't have the ability to "assign" issues to volunteer contributors, and conversely, except for Code of Conduct violations, I don't have any ability to "block" anyone from submitting PRs, either. Perhaps you can both collaborate on this issue? (Or collaborate on two issues)

Regardless, I believe this relevant file is here:

https://github.com/bokeh/bokeh/blob/master/bokehjs/src/lib/models/expressions/stack.ts

that has the function that sums up the values across columns. It occurs to me now that potentially there could be a Boolean property flag on the `Stack` models that toggles two modes:
 
* nan and null both treated as nan
* nan and null both treated as zero

That would only require a small change in this one place, provide the most flexibility for stacking and what I would suggest.

**Comment by Hemansh31** [comment]
@bryevdv  I have tried some changes in the code and tried to build the BokehJS post that according to the instructions in the guide but unfortunately my changes are not getting reflected. A little help please!!!

**Comment by bryevdv** [comment]
If you've made local changes to BokehJS, then you will want to use "inline" resources, so that th local version of BokehJS with the changes is actually used when you create output. There are various ways to do that, I usually set the environment variable `BOKEH_RESOURCES=inline`. 

FYI for general dev questions not specifically related to the implementation of this PR, I'd suggest moving to the Discourse "Development" category, to keep the message history here short and clean: https://discourse.bokeh.org/c/development/7

**Comment by arihans** [comment]
@Hemansh31 @Ak-Bhatia are working on the issue ?!

**Comment by xblahoud** [comment]
Are there some plans to address this issue?

**Comment by bryevdv** [comment]
@xblahoud we had hoped someone new would take up this issue, but that has not been the case. I will triage for review in ` branch-3.0` but given there is a mitigation (set nans->zeros before passing to Bokeh) it is not the highest priority. We may also just go the MPL route and start raising an error if there are nans in stackers.

**Comment by xblahoud** [comment]
Thanks @bryevdv 

I believe that at least raising an error would be a nice thing to do. For me, it took some time (and cross-verification with another plot) to find out that I miss half of the data in one column.

The mitigation can be advised in the error message.
